src/realm/object-store/sync/sync_manager.cpp

```diff
@@ -230,10 +230,22 @@ void SyncManager::reset_for_testing()
     }
 
     {
-        util::CheckedLockGuard lock(m_session_mutex);
+        util::CheckedUniqueLock lock(m_session_mutex);
+
+        bool no_sessions = !do_has_existing_sessions();
+        // There's a race between this function and sessions tearing themselves down waiting for m_session_mutex.
+        // So we give up to a 5 second grace period for any sessions being torn down to unregister themselves.
+        auto since_poll_start = [start = std::chrono::steady_clock::now()] {
+            return std::chrono::steady_clock::now() - start;
+        };
+        for (; !no_sessions && since_poll_start() < std::chrono::seconds(5);
+             no_sessions = !do_has_existing_sessions()) {
```

Comment:
This could be written as `bool no_sessions; while ((no_sessions = !do_has_existing_sessions()) && since_poll_start() < std::chrono::seconds(5)) {`, but maybe that's just trying too hard to avoid duplicating a trivial condition.